# SOD_PhD Mini task
1. Why?
    - This exercise will help me learn how to call variants based on a reference genome.
2. Goals?
    - To learn how to map reads, to genotype samples of interest, and to learn the tools necessary for this exercise.
3. Tools?
    - BWA
4. Timeline?
    - The exercise should be started in the next week (before May 20), completion will happen in the next month or so.

# SOD_PhD
A repo for code &amp; other relevant files pertaining to Hazel Daniels' PhD at Oregon State University

1. Why am I doing this?
    - This project (C2) is one portion of my dissertation. It focuses on the genetic relatedness of geographically neighboring infections of the  Sudden Oak Death (SOD) pathogen *Phytophthora ramorum*.

2. What are my goals?
    - The goal of C2 is to determine through whole genome sequencing (performed by CGRB):
      - Whether subsequent SOD infections occur near previously treated sites at a rate greater than chance, and
      - Whether subsequent SOD infections that occur near previously treated sites are more genetically related than those that do not (i.e. sites further away).

3. What tools do I think I may use?
    - No idea! Definitely R, not sure what packages to use though. Taking suggestions. :3
    - Poppr?

4. What is my timeline for completion?
    - Sequencing: June
    - Cleanup: July
    - Analysis: August
